src/library/autodj/autodjprocessor.cpp

```diff
@@ -817,61 +965,204 @@ void AutoDJProcessor::calculateTransition(DeckAttributes* pFromDeck,
+        // When outro > intro:
+        // ------|ooo|ooo
+        //       |iii|------
+        //
+        // When outro < intro:
+        // ------|ooo|
+        //       |iii|iii-----
+        //
+        // If only the outro or intro length is marked but not both, use the one
+        // that is marked for the transition time. If neither is marked, fall
+        // back to the transition time from the spinbox.
+        if (outroLength > 0 && introLength > 0) {
+            pFromDeck->fadeBeginPos = outroStart;
+            if (outroLength > introLength) {
+                // Cut off end of outro
+                pFromDeck->fadeEndPos = outroStart + introLength;
+            } else {
+                pFromDeck->fadeEndPos = outroEnd;
+            }
+            pToDeck->startPos = introStart;
+        } else if (outroLength > 0 && introLength <= 0) {
+            pFromDeck->fadeBeginPos = outroStart;
+            pFromDeck->fadeEndPos = outroEnd;
+            pToDeck->startPos = introStart;
+        } else if (introLength > 0 && outroLength <= 0) {
+            pFromDeck->fadeBeginPos = outroEnd - introLength;
+            pFromDeck->fadeEndPos = outroEnd;
+            pToDeck->startPos = introStart;
         } else {
-            pFromDeck->posThreshold = 1.0;
-            pFromDeck->fadeDuration = 0.0;
+            useFixedFadeTime(pFromDeck, pToDeck,
+                    getLastSoundPosition(pFromDeck),
+                    getFirstSoundPosition(pToDeck));
         }
+        break;
+    case TransitionMode::FixedSkipSilence:
+        useFixedFadeTime(pFromDeck,
+                pToDeck,
+                getLastSoundPosition(pFromDeck),
+                getFirstSoundPosition(pToDeck));
+        break;
+    case TransitionMode::FixedFullTrack:
+    default:
+        useFixedFadeTime(pFromDeck, pToDeck, fromTrackDuration, 0);
+    }
 
-        if (toTrackDuration > 0.0) {
-            // Transition time is negative here, so we need to add instead of subtract.
-            pToDeck->startPos = (toTrackIntroStart + m_transitionTime) / toTrackDuration;
-        } else {
-            pToDeck->startPos = 0.0;
+    // These are expected to be a fraction of the track length.
+    pFromDeck->fadeBeginPos /= fromTrackDuration;
+    pFromDeck->fadeEndPos /= fromTrackDuration;
+    pToDeck->startPos /= toTrackDuration;
+
+    pFromDeck->isFromDeck = true;
+    pToDeck->isFromDeck = false;
+
+    DEBUG_ASSERT(pFromDeck->fadeBeginPos <= 1);
+
+    if (sDebug) {
+        qDebug() << this << pFromDeck->fadeBeginPos << pFromDeck->fadeEndPos
+                << pToDeck->startPos;
+    }
+}
+
+void AutoDJProcessor::useFixedFadeTime(DeckAttributes* pFromDeck,
+        DeckAttributes* pToDeck, double endPoint, double startPoint) {
+    if (m_transitionTime > 0.0) {
+        // Guard against the next track being too short. This transition must finish
+        // before the next transition starts.
+        double toDeckOutroStart = getOutroStartPosition(pToDeck);
```

Comment:
this need to be done transition mode dependent.

Reply:
What do you mean?

Reply:
This code is also used for full track and skip silence mode. In these cases the outro start point might has been already passed.
Thinking about this again it will not introduce a real issue.
So we may fix it or decument why it is ok.

Reply:
> endPoint - m_transitionTime

Yes, that seems more appropriate. I don't think the outro cues should be taken into account here. Beyond that, I don't really understand the logic here.

```
toDeckOutroStart = (end - startPoint) / 2 + startPoint
```
What is the reasoning behind this? Why the division by 2?

Reply:
https://github.com/mixxxdj/mixxx/pull/2103#issuecomment-535760843